tests/thread/host/spinlock_test_host.cpp

```diff
@@ -0,0 +1,158 @@
+    {
+        if (g_flag_withlock)
+        {
+            enc_spin_lock(enclave, &g_lock);
+        }
+
+        enc_malloc(enclave, &buffer, LARGE_ENOUGH_MEMORY_SIZE, &t_errno);
+
+        if (ENOMEM != t_errno)
+        {
+            printf(
+                "[errno_thread1]Error check errno, expected %d, got %d.\n",
+                ENOMEM,
+                t_errno);
+            g_error_num |= 0x8;
+            if (g_flag_withlock)
+            {
+                enc_spin_unlock(enclave, &g_lock);
+            }
+        }
+
+        if (g_flag_withlock)
+        {
+            enc_spin_unlock(enclave, &g_lock);
+        }
+    }
+
+    return NULL;
+}
+
+void* errno_thread2(void* para)
+{
+    int i = 0;
+    long res = 0;
+    int t_errno = 0;
+    oe_enclave_t* enclave = (oe_enclave_t*)para;
+
+    for (i = 0; i < ERRNO_TEST_TIMES; i++)
+    {
+        if (g_flag_withlock)
+        {
+            enc_spin_lock(enclave, &g_lock);
+        }
+
+        char num_str[] = "123456789987654321123456789987654321";
+        enc_strtol(enclave, &res, num_str, NULL, 10, &t_errno);
+
+        if (ERANGE != t_errno)
+        {
+            printf("[errno_thread2] expected %d, errno %d.\n", ERANGE, t_errno);
+            g_error_num |= 0x40;
+            if (g_flag_withlock)
+            {
+                enc_spin_unlock(enclave, &g_lock);
+            }
+        }
+
+        if (g_flag_withlock)
+        {
+            enc_spin_unlock(enclave, &g_lock);
+        }
+    }
+    return NULL;
+}
+
+void test_errno_multi_threads_sameenclave(oe_enclave_t* enclave)
+{
+    printf("test_errno_multi_threads_sameenclave Starting\n");
+    g_flag_withlock = 1;
+
+    std::thread threads[NUM_SPINLOCK_TEST_THREADS];
+    for (size_t i = 0; i < NUM_SPINLOCK_TEST_THREADS; i++)
+    {
+        if (i & 1)
+        {
+            threads[i] = std::thread(errno_thread1, enclave);
+        }
+        else
+        {
+            threads[i] = std::thread(errno_thread2, enclave);
+        }
+    }
+
+    for (size_t i = 0; i < NUM_SPINLOCK_TEST_THREADS; i++)
+    {
+        threads[i].join();
+    }
+
+    g_flag_withlock = 0;
+    printf("test_errno_multi_threads_sameenclave Complete\n");
+
+    OE_TEST(g_error_num == 0);
+}
+
+void test_errno_multi_threads_diffenclave(
+    oe_enclave_t* enclave1,
+    oe_enclave_t* enclave2)
+{
+    printf("test_errno_multi_threads_diffenclave Starting\n");
+    g_flag_withlock = 0;
+    g_error_num = 0;
+
+    std::thread threads[NUM_SPINLOCK_TEST_THREADS];
+
+    for (size_t i = 0; i < NUM_SPINLOCK_TEST_THREADS; i++)
+    {
+        if (i & 1)
+        {
+            threads[i] = std::thread(errno_thread1, enclave1);
+        }
+        else
+        {
+            threads[i] = std::thread(errno_thread2, enclave2);
+        }
+    }
+
+    for (size_t i = 0; i < NUM_SPINLOCK_TEST_THREADS; i++)
+    {
+        threads[i].join();
+    }
+
+    printf("test_errno_multi_threads_diffenclave Complete\n");
```

Comment:
If my understanding is correct, you are testing if you can have different threads calling into different enclaves generating different errnos inside the two different enclaves. Is this test ensuring that the errnos inside different enclaves dont overwrite each other? If it is, since different enclaves have different thread structures, I am not sure I understand why this is needed.

Reply:
yes, you are right. These cases are testing errno is not overwritten by multithread in the same/different enclaves.  These Cases focus on errno testing, i will change file name from "spinlock_test_host.cpp" to "errno_test_host.cpp".